developer-workflow/extension-modules.rst

```diff
@@ -5,13 +5,489 @@
 Standard library extension modules
 ==================================
 
-In this section, we could explain how to write a CPython extension with the C language, but the topic can take a complete book.
+In this section, we are interested in extending the CPython project with
+an :term:`extension module`. We will not explain how to write the module
+in C but rather explain how to configure the project and make it compile.
 
-For this reason, we prefer to give you some links where you can read a very good documentation.
-
-Read the following references:
+For writing an extension module in C, we prefer to give you some links
+where you can read good documentation:
 
 * https://docs.python.org/dev/c-api/
 * https://docs.python.org/dev/extending/
 * :pep:`399`
 * https://pythonextensionpatterns.readthedocs.io/en/latest/
+
+Adding an extension module to CPython
+=====================================
+
+Assume that the standard library contains a pure Python module :mod:`!foo`
+together with the following :func:`!foo.greet` function:
+
+.. code-block:: python
+
+   def greet():
+       return "Hello World!"
+
+Instead of using the Python implementation of :func:`!foo.greet`, we want to
+use its corresponding C implementation exposed in some :mod:`!fastfoo` module
+written in C. Ideally, we want to modify ``foo.py`` as follows:
+
+.. code-block:: python
+
+   try:
+       # use the C implementation if possible
+       from fastfoo import greet
+   except ImportError:
+       # fallback to the pure Python implementation
+       def greet():
+           return "Hello World!"
+
+Some modules in the standard library are implemented both in C and in Python,
+such as :mod:`decimal` or :mod:`itertools`, and the C implementation is expected
+to improve performance when available (such modules are commonly referred
+to as *accelerator modules*). In our example, we need to determine:
```

Comment:
Oh thanks for the PEP tip, I wasn't aware of it! I think the introduction is the correct place to talk about them in their generality.

Reply:
Update: actually, I said that `itertools` was both in C and Python but there is no Python implementation of itertools... Now, I use `datetime` and `pickle` as examples since they really have identical Python implementation.